### PR TITLE
[#105] tf_idf with word2vec

### PR DESCRIPTION
**Pull request by TaimoorRana**
Integrated Word2Vec with TF-IDF

### CODE REVIEW

**Comment by arekmano** [comment]
Does this improve the overall performance of the clustering?

**Review by Samuel-Campbell — APPROVED** [review]

**Comment by Samuel-Campbell** [comment]
Getting money

File: src/ml_service/GlobalVariables/GlobalVariable.py [hunk 1]
```diff
@@ -7,4 +7,4 @@
 
 
 class Global:
-    Precedence_Directory = r"/home/charmander/Data/text_bk/"
+    Precedence_Directory = r"/Users/taimoorrana/Downloads/text_bk/"
```
Comment (arekmano) [inline src/ml_service/GlobalVariables/GlobalVariable.py]: can this be a relative directory? Constantly switching doesn't make sense

**Review by arekmano — CHANGES_REQUESTED** [review]

File: src/ml_service/.gitignore [hunk 2]
```diff
@@ -6,7 +6,8 @@ coverage.xml
 .idea/
 ./**/*.iml
 .idea/*
-*.bin
+non-lem.bin
+/src/ml_service/feature_extraction/Clustering/k_means/cluster_dir/*
```
Comment (dodoels) [inline src/ml_service/.gitignore]: sometimes `cluster_dir` is not generated then throw errors, I am thinking if we should push this empty directory to the repo and add a `.gitignore` file inside

File: src/ml_service/GlobalVariables/GlobalVariable.py (hunk 1 above)
Comment (dodoels) [reply src/ml_service/GlobalVariables/GlobalVariable.py]: make it `ml_service/data` and gitignore it

File: src/ml_service/.gitignore (hunk 2 above)
Comment (TaimoorRana) [reply src/ml_service/.gitignore]: is this throwing an error??

Comment (dodoels) [reply src/ml_service/.gitignore]: @Taimoorrana1 I cannot reproduce now, but it happened before because it doesnt generate `cluster_dir`

**Review by dodoels — APPROVED** [review]

File: src/ml_service/.gitignore (hunk 2 above)
Comment (TaimoorRana) [reply src/ml_service/.gitignore]: I'm surprised, I think git wouldn't care if the directory isn't found because it was going to be ignored either way.

**Comment by dodoels** [comment]
tested with 2k precedents
```
Loading cached Pickle of words.
Loading word vector file... May take a few seconds
Loading complete
TF-IDF set to: False
Reading: 2000 files
Fetching from precedence
Data Extraction: 100.000000 
Preprocessing complete
Word cache saved
Loading information from dictionary into format for clustering
Transforming informaiton into matrix
DBSCAN begin

Clustering time:
213.07607913017273
HDBSCAN begin
Standardizing matrix
Clustering

Clustering time:
609.2210216522217
K-MEANS begin

Clustering time:
58.98891234397888
```

**Review by arekmano — APPROVED** [review]

File: src/ml_service/feature_extraction/Clustering/main.py [hunk 3]
```diff
@@ -34,28 +39,47 @@ def cluster_hdbscan(data_tuple):
 
 
 if __name__ == '__main__':
-    parser = Precedence_Parser(tfidf=False)
-    precedence_dict = parser.parse_files(Global.Precedence_Directory, 10)
+    # set tfidf to true if you want to use it
+    tf = False
+    nb_files = 100
+
```
Comment (Samuel-Campbell) [inline src/ml_service/feature_extraction/Clustering/main.py]: can you add enum/variable for Strings: 'facts' , 'decisions' 
Lets keep all the config in the top here

File: src/ml_service/feature_extraction/Clustering/main.py [hunk 4]
```diff
@@ -34,28 +39,47 @@ def cluster_hdbscan(data_tuple):
...
     related_word_fetcher.save_cache()
+    print('Word cache saved')
 
     X = []
     labels = []
     precedence_files = []
     piped_fact = []
 
+    print('Loading information from dictionary into format for clustering')
     for fact in precedence_dict['facts']:  # replace this variable with 'decisions' for outcomes
```
Comment (Samuel-Campbell) [inline src/ml_service/feature_extraction/Clustering/main.py]: replace the key here with the enum/variable 
precedence_dict['facts'] OR  precedence_dict['decisions']

They must also be updated in the for loop

**Comment by TaimoorRana** [comment]
@Samuel-Campbell Made final changes that were requested. Please have a look, I will merge it in about an hour.

**Review by Samuel-Campbell — APPROVED** [review]

**Comment by Samuel-Campbell** [comment]
looks good

**Review by arekmano — COMMENTED** [review]
Added some comments

File: src/ml_service/WordVectors/FrenchVectors.py [hunk 5]
```diff
@@ -71,9 +87,16 @@ def vectorize_sent(word_list):
         for word in word_list:
             if word in FrenchVectors.custom_stop_words:
                 continue
```
Comment (arekmano) [inline src/ml_service/WordVectors/FrenchVectors.py:89]: What is this for? Why are we iterating?

File: src/ml_service/feature_extraction/Clustering/dbscan/dbscan.py [hunk 6]
```diff
@@ -3,36 +3,45 @@
...
+        os.makedirs(output_directory)
+    unique_labels = set(labels)
+    for label in unique_labels:
+        file = open(output_directory + str(label) + '.txt', 'w')
+
+        for i, sent in enumerate(data_tuple[InformationType.FACTS.value][labels == label]):
+            file.writelines(sent + '\n')  # original sentence
+
+        file.writelines("-------------------------\n")
+        for i, process_sent in enumerate(data_tuple[InformationType.PROCESSED_FACTS.value][labels == label]):
```
Comment (arekmano) [inline src/ml_service/feature_extraction/Clustering/dbscan/dbscan.py:40]: What is the difference between a fact and a processed fact?

File: src/ml_service/feature_extraction/Clustering/main.py [hunk 7]
```diff
@@ -34,28 +39,47 @@ def cluster_hdbscan(data_tuple):
...
-    precedence_dict = parser.parse_files(Global.Precedence_Directory, 10)
+    # set tf-idf to true if you want to use it
+    tf = False
+    nb_files = 100
+    data_to_extract = 'facts'  # replace this variable with 'facts' for facts and 'decisions' for outcomes
+
+    parser = Precedence_Parser(tfidf=tf)
+    print('TF-IDF set to: ' + str(tf))
+    print('Reading: ' + str(nb_files) + " files")
+    precedence_dict = parser.parse_files(Global.Precedence_Directory, nb_files)
```
Comment (arekmano) [inline src/ml_service/feature_extraction/Clustering/main.py:50]: Remove useless print

**Pull request merged**